mbed_targets/_internal/target_attribute_hierarchy_parsers/accumulating_attribute_parser.py

```diff
@@ -0,0 +1,134 @@
+) -> Dict[str, Any]:
+    """Adds an attribute element to an attribute.
+
+    Args:
+        accumulator: a store of attributes to be updated
+        attribute_name: name of the attribute to update
+        elements_to_add: element to add to the attribute list
+
+    Returns:
+        The accumulator object with the new elements added
+    """
+    for element in elements_to_add:
+        accumulator[attribute_name].append(element)
+    return accumulator
+
+
+def _remove_attribute_element(
+    accumulator: Dict[str, Any], attribute_name: str, elements_to_remove: List[Any]
+) -> Dict[str, Any]:
+    """Removes an attribute element from an attribute.
+
+    Args:
+        accumulator: a store of attributes to be updated
+        attribute_name: name of the attribute to update
+        elements_to_remove: element to remove from the attribute list
+
+    Returns:
+        The accumulator object with the desired elements removed
+    """
+    for element in elements_to_remove:
+        accumulator[attribute_name].remove(element)
+    return accumulator
+
+
+def _calculate_attribute_elements(
+    attribute_name: str, starting_state: Dict[str, Any], applicable_accumulation_order: List[dict]
+) -> Dict[str, Any]:
+    """Adds and removes elements for an attribute based on the definitions encountered in the hierarchy.
+
+    Args:
+        attribute_name: name of the attribute to update
+        starting_state: the list of elements that defines the starting state of the attribute
+        applicable_accumulation_order: the targets in the inheritance tree that can contain add and remove modifiers
+
+    Returns:
+        The elements calculated for an attribute according to the target's inheritance
+    """
+    accumulator = starting_state
+    for target in reversed(applicable_accumulation_order):
+        if f"{attribute_name}_add" in target.keys():
+            to_add = target[f"{attribute_name}_add"]
+            _add_attribute_element(accumulator, attribute_name, to_add)
+        if f"{attribute_name}_remove" in target.keys():
+            to_remove = target[f"{attribute_name}_remove"]
+            _remove_attribute_element(accumulator, attribute_name, to_remove)
+    return accumulator
+
+
+def _determine_accumulated_attributes(accumulation_order: List[dict]) -> Dict[str, Any]:
+    """Finds all the accumulated attributes for a target from its list of ancestors.
+
+    Iterates through the order of inheritance (accumulation order) to find the nearest definition
+    of an attribute, then retraces backwards calculating additions and deletions that modify it.
+
+    Args:
+        accumulation_order: the order in which the targets inherit from the target itself to its highest ancestor
+
+    Returns:
+        A dictionary containing all the accumulating attributes for a target
+    """
+    accumulated_attributes = {}
+    for attribute_name in ACCUMULATING_ATTRIBUTES:
+        for target in accumulation_order:
```

Comment:
We could look into using `itertools.product` here instead of a nested `for`.

Reply:
I've had a go at refactoring this. Seems like a bit of an improvement but not entirely sure.

Reply:
Looked at it again and it definitely is an improvement - thanks for the suggestion 👍